### PR TITLE
Fix #8429 ForwardedHeaderTransformer eats `X-Forwarded-*` headers

### LINKED ISSUE #8324: Can not use swagger ui to test entity endpoint
### How to Reproduce
- Login as superuser
- Go to api docs

### Expected behavior
Interactive API docs

### Observed behavior
Error message; 'Failed to load spec.'

### LINKED ISSUE #8375: Fdp response contains invalid URLs
### How to Reproduce
1. Create a user [ fdp ]
2. Create a group [ fdp ]
3. Assign the fdp user to the fdp group as [ manager ]
4. Login a fdp user
5. Import [FDP.xlsx](https://github.com/molgenis/molgenis/files/3052892/FDP.xlsx) (e.g. using the importwizard plugin)
6. Import [FDP_test_data.xlsx](https://github.com/molgenis/molgenis/files/3052893/FDP_test_data.xlsx) (e.g. using the importwizard plugin)
7. Navigate to https://latest.test.molgenis.org/api/fdp/catalogue

### Expected behavior
- Reponse contains URL **https**://latest.test.molgenis.org/api/fdp/catalogue

### Observed behavior
- **http**://latest.test.molgenis.org/api/fdp/catalogue

### LINKED ISSUE #8384: R script executor: Heatmap script can't be executed
### How to Reproduce
See test case SCRIPTSYS004
1. Create a 'data'-group
2. Import the [Rtest.xlsx](https://github.com/molgenis/molgenis/files/3053928/Rtest.xlsx) script in the 'data'-folder from the data group
3. "Create script Heatmap

Name: Heatmap
Type: R
Content: 

  library(RCurl)
  eval(expr = parse(text = getURL(""https://latest.test.molgenis.org/molgenis.R?molgenis-token=${molgenisToken}"")))
  df <- molgenis.get(""data_Rtest"")
  df2 <- apply(df, 1,as.numeric)
  is.na(df2) <- sapply(df2, is.infinite)
  df2[is.na(df2)] <- 0
  df2[is.nan(df2)] <- 0
  png('${outputFile}')
  heatmap(as.matrix(df2),symm=FALSE,trace=""none"")
  dev.off()
  molgenis.logout()

Generate security token: Yes
Result file extension: png
"

### Expected behavior
Script runs succesfully

### Observed behavior
1) Test fails
```
10:52:25.029 - Execution started.
10:52:25.955 - Failed. no lines available in input

In call:
read.table(file = file, header = header, sep = sep, quote = quote, 
    dec = dec, fill = fill, comment.char = comment.char, ...)
```
2) Timestamps incorrect

### LINKED ISSUE #8429: ForwardedHeaderTransformer eats `X-Forwarded-*` headers
The Javadoc of `ServletUriComponentBuilder` is misleading and outdated (see https://github.com/spring-projects/spring-framework/issues/21850). Instead of `ForwardedHeaderTransformer` a `ForwardedHeaderFilter` should be used to retain the `X-Forwarded-*` headers.

Related issues:
#8384 R script executor: Heatmap script can't be executed
#8324 Can not use swagger ui to test entity endpoint
#8375 FDP response contains invalid URLs

### PR DESCRIPTION
**Pull request by tommydeboer**
Fixes #8384 R script executor: Heatmap script can't be executed
Fixes #8324 Can not use swagger ui to test entity endpoint
Fixes #8375 FDP response contains invalid URLs

#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- User documentation updated
- (If you have changed REST API interface) view-swagger.ftl updated
- Test plan template updated
- [x] Clean commits
- [x] Added Feature/Fix to release notes

### CODE REVIEW

File: molgenis-core-ui/src/main/java/org/molgenis/core/ui/MolgenisWebAppInitializer.java [hunk 1]
```diff
@@ -58,6 +59,12 @@ protected void onStartup(ServletContext servletContext, Class<?> appConfig, int
     }
 
     // add filters
+    Dynamic forwardedHeaderFilter =
+        servletContext.addFilter("forwardedHeaderFilter", ForwardedHeaderFilter.class);
+    forwardedHeaderFilter.setAsyncSupported(true);
+    forwardedHeaderFilter.addMappingForUrlPatterns(
+        EnumSet.of(DispatcherType.REQUEST, DispatcherType.ASYNC), false, "*");
```
Comment (dennishendriksen) [inline molgenis-core-ui/src/main/java/org/molgenis/core/ui/MolgenisWebAppInitializer.java]: What happens if I create an endpoint ```/test``` that returns ```forward:/test-forwarded``` that returns the current request URL. This determines whether you need to add DispatcherType.FORWARD to this list (and maybe others have well?).

File: molgenis-security/src/main/java/org/molgenis/security/account/AccountController.java [hunk 2]
```diff
@@ -150,18 +149,10 @@ public String changePasswordToken(
...
-                .replacePath(URI + "/activate")
-                .build()
-                .toUriString();
-      } else {
-        String scheme = request.getHeader("X-Forwarded-Proto");
-        if (scheme == null) scheme = request.getScheme();
-        activationUri = scheme + "://" + request.getHeader("X-Forwarded-Host") + URI + "/activate";
-      }
+      String activationUri =
+          ServletUriComponentsBuilder.fromCurrentRequest()
```
Comment (dennishendriksen) [inline molgenis-security/src/main/java/org/molgenis/security/account/AccountController.java]: Might the ```ServletUriComponentsBuilderFactory``` bean be useful in cases such as these, e.g. to simplify testing?

File: molgenis-security/src/test/java/org/molgenis/security/account/AccountControllerTest.java [hunk 3]
```diff
@@ -120,69 +118,6 @@ public void activateUserMolgenisUserException() throws Exception {
...
-            content()
-                .string(
-                    "{\"message\":\""
-                        + AccountController.REGISTRATION_SUCCESS_MESSAGE_USER
-                        + "\"}"));
-    ArgumentCaptor<User> molgenisUserCaptor = ArgumentCaptor.forClass(User.class);
-    ArgumentCaptor<String> baseActivationUriCaptor = ArgumentCaptor.forClass(String.class);
-    verify(accountService)
-        .createUser(molgenisUserCaptor.capture(), baseActivationUriCaptor.capture());
-    assertEquals(baseActivationUriCaptor.getValue(), "http://website.com/account/activate");
```
Comment (dennishendriksen) [inline molgenis-security/src/test/java/org/molgenis/security/account/AccountControllerTest.java]: I think we should test whether the URLs that we generate are correct

File: molgenis-core-ui/src/main/java/org/molgenis/core/ui/MolgenisWebAppInitializer.java (hunk 1 above)
Comment (tommydeboer) [reply molgenis-core-ui/src/main/java/org/molgenis/core/ui/MolgenisWebAppInitializer.java]: Tested like this:
```java
  @GetMapping(value = "/test")
  public ModelAndView forward() {
    return new ModelAndView(format("forward:%s/test-forwarded", BASE_URI));
  }

  @GetMapping(value = "/test-forwarded")
  @ResponseBody
  public String forwarded(HttpServletRequest request) {
    return ServletUriComponentsBuilder.fromCurrentRequest().toUriString();
  }
```

Doing a request with `X-Forwarded-Proto:https` gives: `"http://localhost:8080/api/v1/test-forwarded"`

File: molgenis-security/src/main/java/org/molgenis/security/account/AccountController.java (hunk 2 above)
Comment (tommydeboer) [reply molgenis-security/src/main/java/org/molgenis/security/account/AccountController.java]: This factory currently sits in the `molgenis-api-data` module. Do you have a suggestion on where to put it? 

`molgenis-api-data` and  `molgenis-security` both share `molgenis-data` as a dependency but that doesn't sound like a nice fit to me.

**Comment by tommydeboer** [comment]
**Attention:**
Testing this PR on a preview server has brought to light other problems that are related to deployment: the port is not passed in a header.

--------

**Swagger**
The Swagger plugin doesn't work. Error:
```
"Can't read from file https://preview-pr-8434-2.dev.molgenis.org:80/plugin/swagger/swagger.yml"
```

Note that port `80` has been appended to the URI. The protocol (`https`) has been correctly set.

---------

**R client**
The R client now includes the correct protocol but also the wrong port. This can be further verified by doing the following requests:

---------
`GET https://preview-pr-8434-2.dev.molgenis.org/molgenis.R`
Headers:
`X-Forwarded-Proto:https`

Results in:
```
    molgenis.api.url <- paste0("https://preview-pr-8434-2.dev.molgenis.org:80/molgenis.R/api/",'v1/')
    molgenis.api.url.v2 <- paste0("https://preview-pr-8434-2.dev.molgenis.org:80/molgenis.R/api/",'v2/')
```

--------

`GET https://preview-pr-8434-2.dev.molgenis.org/molgenis.R`
Headers:
`X-Forwarded-Proto:https`
`X-Forwarded-Port:443` (this is the actual port)

Results in:
```
    molgenis.api.url <- paste0("https://preview-pr-8434-2.dev.molgenis.org/molgenis.R/api/",'v1/')
    molgenis.api.url.v2 <- paste0("https://preview-pr-8434-2.dev.molgenis.org/molgenis.R/api/",'v2/')
```

**Comment by tommydeboer** [comment]
Confirmed that the FDP endpoint now also shows the correct protocol

File: molgenis-security/src/main/java/org/molgenis/security/account/AccountController.java (hunk 2 above)
Comment (tommydeboer) [reply molgenis-security/src/main/java/org/molgenis/security/account/AccountController.java]: Using the factory introduces a lot of complexity in the tests so we'll leave it as is

File: molgenis-security/src/test/java/org/molgenis/security/account/AccountControllerTest.java (hunk 3 above)
Comment (tommydeboer) [reply molgenis-security/src/test/java/org/molgenis/security/account/AccountControllerTest.java]: Tested in the API test

**Pull request merged**